### ISSUE TITLE
[question] Using Ninja (Multi-Config) with Extending Own CMake Presets Example 

### ISSUE THREAD

**Issue by wayneph01** [conan-io/conan#13968]
### What is your question?

Howdy, I am having troubles getting one of the CMake examples working with Ninja, and I assume I am doing something wrong, but can't for the life of me figure out what.

I have cloned (pulled changes this morning) the examples2 repo and am starting with the extend_own_cmake_presets example.  For my tests, I am just trying to specify the generator in conanfile.py on line 29.  When attempting to use `tc = CMakeToolchain(self, generator="Ninja Multi-Config")`, I am consistently getting bad CMakePresets which doesn't let me continue.

- On Both Windows and Linux, if I run `cmake --list-prests` before installing, I get the expected "can't read ConanPresets.json" since the referenced file isn't created yet.
- Next, I ran `conan install .` on both operating systems and running `cmake --list-presets` again, I got errors of "Duplicate Presets" on both Operating systems, however they did different things under the covers.  For both systems, ConanPresets.json contains a "duplicate conan-debug" buildPreset.  On Linux, it appeared to use a single config template as the CMakePresets.json file that was included is in the "build/Release/generators" directory as opposed to the "build/generators" directory that was created on Windows.
- Next I ran `conan install . -s build_type=Debug` on both systems.  Windows has appeared to correct the previous "Duplicate" presets warning as `cmake --list-presets` now lists all 6 presets.  However, on linux I'm still getting a "Duplicate Presets" error.  The duplicate buildPreset in ConanPresets.json has been cleared, up but now there are two includes from the build/Release and build/Debug directories that both define different versions of the conan-default configurePreset.

I blew away ConanPresets.json and the build directory to try the regular non Multi-Config version of Ninja using `tc = CMakeToolchain(self, generator="Ninja")`

- I started again by running `conan install .` to create ConanPresets.json.  It appears the "exact" same files were generated as before, with Windows now using the Multi-Config option even though just "Ninja" was requested.
- I now ran `conan install . -s build_type=Debug` to generate the other configuration and the results again were reversed.  On Linux, the ConanPresets.json duplicate was resolved and including the two files correctly resolved the rest of the conan- presets being inherited from the main file.  On Windows, ConanPresets was also resolved, but generated CMakePresets.json file in build/generators REPLACED the original Release configuration with Debug configuration causing `cmake --list-presets` to report an "Invalid preset" since none of the files have conan-release that we are trying to inherit from.

I'm not sure if I'm doing something wrong or if I'm running into a bug of some sort.  Even though the generator I pass into the CMakeToolchain call is showing up in the Preset files, I don't think it is actually being fully honored/evaluated when trying to generate the files.

Also, is it the expectation that I would need to install ALL configurations before I can configure/build one of them?  For example, once I add "RelWithDebInfo" to my CMakePresets, it seems like I would have to install it, otherwise I'll keep running into the issue of a "missing" preset since nothing generated conan-relwithdebinfo?

Thanks!

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @wayneph01 

Thanks for your detailed report.

It is possible that either the example is not prepared or documented for the ``Ninja`` generator, or there might still be some gap  in the ``CMakePresets`` generated files.

Just to make sure, I guess that you tried first with the "canonical" generators (Unix Makefiles and Visual Studio in Linux and Windows), and everything worked fine, and it was only when you tried to use Ninja?

**Comment by wayneph01** [comment]
OK, I just went back and tested "as is" and I am seeing the same problem with the canonical generators as I am with Ninja with regards to duplicates.  Once I run `conan install` for all of the defined build types, I am able to successfully run `cmake --list-presets` on both systems.  Since I was planning to use Ninja, I hadn't actually tested the example before starting to make changes.

As another test, I also added the "relwithdebinfo" configPreset and the "multi-relwithdebinfo" and "relwithdebinfo" buildPresets.  With those in there, the ConanPresets.json ended up getting "duplicates" for both "conan-debug" and "conan-relwithdebinfo" when I ran the first `conan install` for Release.  It LOOKS LIKE, the configPresets and buildPresets being created in the ConanPresets.json file are the attempt to make sure there is valid configuration as early as possible to address my 2nd question.  As long as they aren't duplicated, it looks like I SHOULD be able to run `cmake --preset release` after just the first install.

I think I found one more issue (probably with the test, not with Conan itself) on Windows.  Once I successfully installed all of the different variations, I actually ran `cmake --preset release`.  All of the new files ended up in the "main" directory next to the CMakeLists.txt document.  I generally use Ninja, so it may be expected for Visual Studio, but I would have expected them to be somewhere inside of the build directory.  Linux correctly used the build/Release directory for the Unix Makefiles.

**Comment by wayneph01** [comment]
Just started looking into this a little more, and I think I figured out (at least part of) what is going on.

For the duplicate build presets, the code is adding every instance of an "inherit" that it finds in the main CMakePresets.json file.  In the example, there are multiple build presets that inherit from the same 'conan-debug' preset, so it is creating it multiple times.  I think it just needs to make sure it only adds unique items, and it should be good to go.  As a really bad coincidence, my Presets file is configured similarly, so I hadn't noticed that it was directly related to what is found in the file.

I haven't had much time to start looking at any of the generator stuff, but figured I should at least report this much back, as it will lead to an issue in the base product that needs to be fixed.

**Comment by wayneph01** [comment]
Also, still testing, but I think I figured out the main issue with switching from canonical to Ninja Multi-Config.  In short, I was setting the generator in the CMakeToolchain, but NOT the cmake_layout() call.  So the initial directory setup was being figured out using canonical, but the Toolchain had the one I requested.  Things look a lot better now that I'm setting it in both.  I still need to verify that everything is working right, but  I think I at least have a chance.  Instead of specifying the Generator in each recipe, I think I'm just going to add `tools.cmake.cmaketoolchain:generator` to my profiles which would allow us to use different things depending on individual needs. 

@memsharded I know the title isn't as accurate anymore, but can this issue be used to track the "duplicate preset generation" issue, or should I create something a little more specific.  Unless I find something else today, I the generators switching is working ok...

**Comment by wayneph01** [comment]
Sample change I was testing for the duplicate issue in presets.py _collect_user_inherits(...):
From:
```python
               for preset_type in types:
                    for preset in user_json.get(preset_type, []):
                        inherits = preset.get("inherits", [])
                        if isinstance(inherits, str):
                            inherits = [inherits]
                        conan_inherits = [i for i in inherits if i.startswith(preset_prefix)]
                        if conan_inherits:
                            collected_targets.setdefault(preset_type, []).extend(conan_inherits)
        return collected_targets
```
To:
```python
               for preset_type in types:
                    conan_inherits = []
                    for preset in user_json.get(preset_type, []):
                        inherits = preset.get("inherits", [])
                        if isinstance(inherits, str):
                            inherits = [inherits]
                        conan_inherits.extend([i for i in inherits if i.startswith(preset_prefix)])
                    if len(conan_inherits):
                        collected_targets.setdefault(preset_type, []).extend(list(set(conan_inherits)))
        return collected_targets
```
However, now I'm also getting another issue where the "temporary" buildPreset that is being put in the ConanPresets.json file doesn't specify a configurePreset (nor inherit from something that does) so it is still an invalid preset until I have called `conan install` for all build_types.  That is going to be a little more interesting to figure out how to "auto link".

Edited for formatting.

**Comment by wayneph01** [comment]
After more testing, I realized I could fix the remaining problem simply by NOT inheriting from the "conan-" configurations in the buildPresets.  Since all the build preset was doing (once generated) is linking it to a configuration preset, I don't need to use them since all of my presets already specify one of my configurations that inherits from "conan-" already.  It looks like cleaning up the duplicates will be enough of a fix.

I can attempt to make a PR with the fix indicated above so that this can get included in an official release.  Is there anything in the way of unit testing you know of that may also need to be updated.  I haven't had any extra time to try and look into the testing.

**Comment by memsharded** [comment]
Hi @wayneph01,

Thanks for all the feedback, sorry I didn't have time to look into this further.
Seems the fix to avoid duplicates is good. I'd definitely prefer to start with a red test, but don't worry too much about it, if you fire the PR with the fix, that will help us to try to prioritize and we could contribute the test ourselves. I'd probably take https://github.com/conan-io/conan/blob/13c2acdab1fcc142b932c348c46bb13c194c2878/conans/test/integration/toolchains/cmake/test_cmaketoolchain.py#L409 integration test as a start.

**Comment by wayneph01** [comment]
Hopefully I've got everything right for the first submittal.  PR is created, and I took a stab at a unit test related to the updated functionality.  Let me know if you have any questions/issues.

**Comment by memsharded** [comment]
Thanks for contributing it, we will check the PR soon, consider it for next release.

**Comment by memsharded** [comment]
Fixed in https://github.com/conan-io/conan/pull/14296 for next 2.0.9
